library/common/main_interface.h

```diff
@@ -1,12 +1,223 @@
 #pragma once
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
 
 // NOLINT(namespace-envoy)
 
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef int envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP request. Valid only for the duration of the request and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_request_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP stream. Valid only for the duration of the stream and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_stream_t;
+
+/**
+ * Result codes returned by all calls made to this interface.
+ */
+typedef enum { ENVOY_SUCCESS, ENVOY_FAILURE } envoy_status_t;
+
+/**
+ * Error code associated with terminal status of a HTTP stream or request.
+ */
+typedef enum {} envoy_error_code_t;
+
+/**
+ * Holds data about an HTTP request.
+ */
+typedef struct {
+  // Status of the Envoy HTTP request. Note that the request might have failed inline.
+  // Thus the status should be checked before pursuing other operations on the request.
+  envoy_status_t status;
+  // Handle to the Envoy HTTP request.
+  envoy_request_t request;
+} envoy_request;
+
+/**
+ * Holds data about an HTTP stream.
+ */
+typedef struct {
+  // Status of the Envoy HTTP stream. Note that the stream might have failed inline.
+  // Thus the status should be checked before pursuing other operations on the stream.
+  envoy_status_t status;
+  // Handle to the Envoy HTTP Stream.
+  envoy_stream_t stream;
+} envoy_stream;
+
+/**
+ * Holds a single name/value header.
+ */
+typedef struct {
+  char* name;
```

Comment:
Strong preference to create a small `string_view` struct which has a pointer and length and use that everywhere. We should not rely on null termination and have run into problems with this in the main Envoy code (it is now removed).

Reply:
I agree with this sentiment, will definitely make this a bit more ergonomic

Reply:
good idea. changed everywhere we had string-like constructs.